File: src/courses/data/catalog/02-dsa/dsa-core/implementations/csharp/heap.cs
/**
 * C# Implementation: Heap & Priority Queue
 * Invariant: Insert O(log N), ExtractMin O(log N), Peek O(1)
 */
using System;
using System.Collections.Generic;

namespace Seed.DSA.Core
{
    public class HeapPriorityQueue<T>
    {
        private readonly List<T> _items = new List<T>();

        // TODO: Implement algorithm operations
        public void Add(T item)
        {
            _items.Add(item);
        }

        public int Count => _items.Count;
    }
}

File: src/courses/data/catalog/02-dsa/dsa-core/implementations/python/heap.py
"""
Python Implementation: Heap & Priority Queue
Invariant: Insert O(log N), ExtractMin O(log N), Peek O(1)
"""
from typing import Generic, TypeVar, List, Optional

T = TypeVar('T')

class HeapPriorityQueue(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    # TODO: Implement core operational method
    def insert(self, value: T) -> None:
        self._items.append(value)

    def __len__(self) -> int:
        return len(self._items)
